flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActorTest.java

```diff
@@ -395,6 +396,43 @@ public void testOnStartFails() throws Exception {
 		}
 	}
 
+	/**
+	 * Tests that multiple termination calls won't trigger the onStop action multiple times.
+	 * See FLINK-16703.
+	 */
+	@Test
+	public void callsOnStopOnlyOnce() throws Exception {
+		final CompletableFuture<Void> onStopFuture = new CompletableFuture<>();
+		final OnStopCountingRpcEndpoint endpoint = new OnStopCountingRpcEndpoint(akkaRpcService, onStopFuture);
+
+		try {
+			endpoint.start();
+
+			final AkkaBasedEndpoint selfGateway = endpoint.getSelfGateway(AkkaBasedEndpoint.class);
+
+			// try to terminate the actor twice
+			selfGateway.getActorRef().tell(ControlMessages.TERMINATE, ActorRef.noSender());
+			selfGateway.getActorRef().tell(ControlMessages.TERMINATE, ActorRef.noSender());
+
+			// wait a bit so that both terminate messages can be processed
+			try {
+				endpoint.getTerminationFuture().get(10L, TimeUnit.MILLISECONDS);
```

Comment:
Is there no better way to drain the messages?

Reply:
I'll introduce a `CountDownLatch` in the `OnStopCountingRpcEndpoint#onStop` method on which I will wait.

Reply:
ok